### ISSUE TITLE
Tech: Centralize all user settings in single service that also handles cascading.

### ISSUE THREAD

**Issue by Richargh** [MaibornWolff/codecharta#178]
# Feature Request

## Description

This is a tech story with no value for the user, but still important for future features. 

The single service centralizes all settings a user can set. Modules are notified of changes using something like rx.js. 

Cascading means that settings can overwrite each other if they are from different sources. F.ex. the default settings can be overwritten by url parameters.

## TODO
* [x] map is immutable, take map out of settings, markingPackage in settings instead in map
* [x] Simplify settings to only handle getting and updating settings from partial settings objects
* [x] Updated `settings` object, which uses three sub-objects. See interface [codeCharta.model.ts#L32](https://github.com/MaibornWolff/codecharta/blob/feature/178/centralize-settings/visualization/app/codeCharta/codeCharta.model.ts#L32)
* [x] Create `codeCharta.model.ts` file which holds all globally needed interfaces
* [x] Create `codeCharta.service.ts` file which combines simplified functionality from former `codeCharta.component`, `data.service`, `data.validator.service`, `data.loading.service`. The `codeCharta.service.ts` get called from `codeCharta.component` and is supposed to be the global connection to the core functionality like onLoadApplication
* [x] Deleted `data.service`, `data.validator.service`, `data.loading.service`
* [x] Implement ScenarioService behaviour onLoad and onFileImport(?)
* [x] Merge fileSettings from visible `CCFiles` into `settingsService.settings.fileSettings`
* [x] Check if #352 is fixed
* [x] Implement UrlUtils which updates the settings. Don't forget about the `mode=Delta`
* [x] Update content of the dialog when clicking `URL Parameters` button inside sidebar. Only show file and mode keys
* [x] Remove `URL Parameters` functionality from `UrlLUtils` and `SettingsService`, but keep option to import files and set a `mode` (renderState) from within the url
* [x] Adapt the new settings/fileState/metric behaviour step by step to the UI components
* [x] Adapt all tests and try to follow the scheme for setting all objects like [nodeContextMenu.component.spec.ts](https://github.com/MaibornWolff/codecharta/blob/feature/178/centralize-settings/visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.spec.ts)
* [x] Everywhere -> Remove unused injections
* [x] scenarioHelper -> Refactor getScenarioByName array[string]
* [x] Tests -> update skipped Tests 
* [x] settingsService -> onCameraChanged
* [x] fix latency issue with gitignore search/blacklist

**Comment by alschmut** [comment]
According to the ideas @mrpatpat and I had this morning and many further ideas. What do you think about this guys? VariableNames aren't fixed. Feel free to suggest changes!

<img width="80%" src="https://user-images.githubusercontent.com/23529226/51918847-31fcbb00-23da-11e9-85a2-d067eaaa0d87.jpeg">

- The Application CodeCharta itself holds all the imported files and one complete set of all changeable Settings
- A file holds the actual `treeMap`, can have some `meta`-data and **can** but **does not have to** hold a few selected settings (like `blacklist`, `edges`, `colorMetric` etc. could make sense). Those settings never get rendered directly. They will only override the `Application.settings` `onload` or `onChangeSelectedFile`
- The settings include all settings
    - needs a `renderMode`, so the renderingComponents know which maps get rendered how
    - all colors mentioned in `renderSettings`
    - all lists regarding the map like `blacklist`, `attributeTypes`, `edges`
    - all other settings, which are already inside the settingsObject
    - Settings do **NOT** include the map

### Ideas for fileChooser behaviour
- Onload
    - DefaultSettings get applied to `Application.settings`
- When selecting another file from `fileChooser`
    1. All current `Application.settings` keep its values
    1. **IF** the new file includes some `File.settings`, only those get override in `Application.settings`
- When selecting multiple or delta files in `fileChooser` 
    1. All current `Application.settings` keep its values
    1. Some `Application.settings`-objects should get reset to its default values (`focusedNodePath`, `searchPattern`, `searchedNodePaths `, `edges`, `blacklist`)
    1. Some `File.settings`-objects might get merged and converted with new paths (`edges`, `blacklist`) and then set into `Application.settings`
    1. If some metric values clash, use its corresponding defaultSetting values

**Comment by Richargh** [comment]
Looks interesting. Here's a couple of thoughts:

* Primarily, is this turning into a new feature or is it still just redesign?
* What about special `user settings` that we might store only using the NW.js standalone? We don't have that but we might at some nebulous point in the future. Useful for "recent files" or user scenarios.
* Generally it might make sense to think of settings from many views. User Settings, Application Settings (f.ex. defaults), File Settings and Url Settings. And they don't include the same elements. Rather they form more of a venn diagram.
* What about `scenarios` for area, height, color? Is that another setting category?
* At some nebulous point Meta might also include CreationDate

**Comment by alschmut** [comment]
Thanks @Richargh 😉
- This redesign - especially taking out the map of the settings - enables a clearer focus on what the settingsService is actually made for: Updating and saving settings (only), which the user might want to change. @mrpatpat mentioned that **Rx.js** might be a good option to use, as it handles some functionality, we implemented manually, out-of-the-box. But before we could possibly use that, we need to make sure, the structure of data-flow (and where its stored) is a bit clearer.
- Do you think of any `user settings`, which should be outside the general `Settings`-object? What is "recent files"? Maybe the `File.settings` are what you call `user settings`? Those would hold, whatever the user wants to/can export or import
- We/I thought of settings in the following way:
  - `Application.settings`: only these settings are used for current rendered map
  - `DefaultSettings`: MUST hold ALL possible settings to apply onload to `Application.settings`
  - `Scenario` and `File.settings`, `UrlParamSettings`: Are part of `Application.settings`. So they can hold all, some or none those settings. Yea think of this as a Venn diagram (circle inside the circle)

**Comment by alschmut** [comment]
btw just updated [my huge comment](https://github.com/MaibornWolff/codecharta/issues/178#issuecomment-458582079) with some new graphics 😁

**Comment by Richargh** [comment]
* Going reactive is an interesting choice. Should work well with Typescript async/await.
* I think of settings the way CSS thinks about styles. It goes from general to specific:, global -> tags -> classes -> ids. I probably missed some CSS stuff in there. The more specific CSS style overwrites the more general style. 
* What does that mean for our settings, they also go from general to specific: default -> application -> user -> file. That's my initial ordering at least. The more specific one overwrites the more general one. This would mean file overwrites user, which does sound strange, now that I have written it. Something to ponder I guess. 
* Unlike CSS our settings have things that do not make sense to specify on all levels. 
** A user can have a list of "recent files" he openend in the standalone. That only makes sense to store in the user settings. A user might also prefer to view the CodeMap by default from the top. That to me are `user settings`. You have them in almost all programs and they are usually stored in something like `user.properties`.

... I would go on but I think my sentences only go longer and less coherent. I'll sleep about it and talk some more in two weeks :) If you have a solution by then it's fine by me :D

**Comment by alschmut** [comment]
- I think your idea of "overriding settings in a hierarchical order" makes sense and doesn't contradict with the graphs in general.Just changing the order makes probably sense. And as the `defaultSettings` actually are the `Application.settings`,  we can just remove one of those in the list:
  > DefaultSettings + File.settings + currentUser.settings => Application.renderSettings

  The currentUser is in my opinion the user who is using the application in the browser or the standalone. He can change some settings from the UI and sees live updates.

- I definitely agree that `recentlyOpenedFiles` only make sense for the standalone and we should create a separate SettingsObject only for the standalone, when time comes.

- I actually got a good idea: What if your general `user settings`, which fit the current users needs, are called Scenarios. We actually got the backend interface already with PR #335. We just need to do some work, so the user can create those Scenarios on its own #226, export them and import them and for the standalone they should be imported from the users filesystem automatically.

**Comment by alschmut** [comment]
I think we should also separate Settings into two sub-objects: That actually has a similar origin than the `user settings` @Richargh mentioned. Some settings have no connection to the map and some are directly related to the nodes or metrics. The settings, which are not related to the map, are kinda something like `user settings`. And those related to the map, all have to be reset, when changing the RenderMode and showing different (merged) cc.json files. Instead of resetting those manually, we could just say "reset all settings inside the map-related settings-object"... What do you think about that?
<img width="80%" src="https://user-images.githubusercontent.com/23529226/52177727-6fd45780-27bd-11e9-825e-5ca483737730.jpeg">

**Comment by alschmut** [comment]
I just thought a bit about the cc.json, and while writing a short sample on #350, I felt like, we should split the settings even in 3 objects:
- abc (still needs a good name): for actual analysis data which enhances the map data
    - data only comes from cc.json file
    - must not be part of `Scenario.settings`
- mapSettings: settings that are kinda connected to the provided map data
    - must be updated when the shown map changes
    - can be part of `Scenario.settings`
- appSettings: settings which do not have any direct connection to the map
    - can be part of `Scenario.settings`

What do you think about this?
<img width="40%" src="https://user-images.githubusercontent.com/23529226/53201745-af53ce00-361c-11e9-8a33-e8b8433d6d30.jpeg">

**Comment by alschmut** [comment]
See updated main description.

@mrpatpat regarding the UI components and html-templates. As we don't gonna update the settingsService object directly. Do we create viewModel objects everywhere and call a local `updateSettings()` method everywhere? Should we maybe even add such an `updateSettins()` method to the `SettingServiceSubscriber`, because we sometimes have these methods and we call them different everywhere like `onChange()`, `apply()`, `update()`?

**Comment by alschmut** [comment]
I just added some viewModels for a few UI components and feel like there must be a better way to update the settingsObject. Until now we always used one single update/notify method inside every UI component which updated the complete setting object. With the new settingsService architecture, we intend to only do partial updates. Is it worth to write multiple methods, which get called from the html-template and only update the needed settings?

I am thinking about the difference between:
```
// having only one function inside each UI component too update all used settings
applySettings() { 
    const update = { ...allTheSettingsUsedInThisUIComponent}
    this.settingsService.updateSettings(update)
}
```

```
// having multiple functions inside every UI component for every single partial update option
applySettingsAreaMetric() { 
    const update = { ...newAreaMetric}
    this.settingsService.updateSettings(update)
}
applySettingsColorMetric() { 
    const update = { ...newColorMetric}
    this.settingsService.updateSettings(update)
}
...
```

**Comment by mrpatpat** [comment]
> I just added some viewModels for a few UI components and feel like there must be a better way to update the settingsObject. Until now we always used one single update/notify method inside every UI component which updated the complete setting object. With the new settingsService architecture, we intend to only do partial updates. Is it worth to write multiple methods, which get called from the html-template and only update the needed settings?
> 
> I am thinking about the difference between:
> 
> ```
> // having only one function inside each UI component too update all used settings
> applySettings() { 
>     const update = { ...allTheSettingsUsedInThisUIComponent}
>     this.settingsService.updateSettings(update)
> }
> ```
> ```
> // having multiple functions inside every UI component for every single partial update option
> applySettingsAreaMetric() { 
>     const update = { ...newAreaMetric}
>     this.settingsService.updateSettings(update)
> }
> applySettingsColorMetric() { 
>     const update = { ...newColorMetric}
>     this.settingsService.updateSettings(update)
> }
> ...
> ```

I think this should be decided for each case individually e.g. sometimes you need separate methods because the component needs update things separately 



> See updated main description.
> 
> @mrpatpat regarding the UI components and html-templates. As we don't gonna update the settingsService object directly. Do we create viewModel objects everywhere and call a local `updateSettings()` method everywhere? Should we maybe even add such an `updateSettins()` method to the `SettingServiceSubscriber`, because we sometimes have these methods and we call them different everywhere like `onChange()`, `apply()`, `update()`?

I personally prefer the ViewModel solution because it decouples our services from the html bindings. we then know that refactoring a template wont affect core code. We often accessed the settings directly from the template, which was a bad practice anyway (decoupling, law of demeter, ). 
I disagree with adding updateSettings to the subscriber interface since it is not the purpose of this interface. The only thing I can imagine is to make a base class for settings dependent services which implements such an update method. But then again, this introduces some complexity. 

In summary I think we do not need an abstraction layer and calling the update method in components is fine.

**Comment by alschmut** [comment]
@mrpatpat Alright, I know it wasn't good coding practice to not use a viewModel, but it was definitely less code 😄 nevertheless we should use them now. And I gonna implement (without interface then) an `MyComponent.applySettings()` method which calls the `settingsService.updateSettings(update)`

**Comment by alschmut** [comment]
Should the viewModel object hold everything that is accessed in the UI or only those, which are updatable Settings by the UI?

**Comment by Richargh** [comment]
Not sure if it fits your case @alschmut but I would say the ViewModel holds everything that is one-way or two-way binded by the view/UI. That way you can keep your view as stupid as possible and only let it bind to your ViewModel. The ViewModel contains all logic to transfer your model into a format that the view can bind to. 

That gives you the benefit that you can use the ViewModel in your test and don't even have to create the full UI.

**Comment by alschmut** [comment]
Thanks @Richargh! I agree and have to admit my question wasn't put right. The question is rather, if other objects (not coming from the settings), which are needed in the view, should be stored inside the viewModel as well or separate

**Comment by Richargh** [comment]
Not sure I read your question right. If your view is divided into multiple cohesive areas, one for settings, one for other stuff, then this answer is for you: One ViewModel per cohesive area. So Settings gets one ViewModel and the other non-settings stuff gets another ViewModel or more.

**Comment by Richargh** [comment]
Url Parameter can be removed. The only useful feature is opening specific files in normal, multiple or delta mode.
